### PR TITLE
Extend OpenAI Python Interface to Async Version for Promptflow

### PR DESCRIPTION
**Pull request by liucheng-ms**
# Description

This pull request introduces asynchronous functionality to the OpenAI Python interface within Promptflow. The aim is to support asynchronous Python code, which is crucial for efficient handling of concurrent tasks and improving overall performance. 

We have extended our existing decorators that inject Promptflow's customized logic when invoking the OpenAI Python interface to operate asynchronously. This enhancement aligns with Promptflow's goal of providing a seamless and efficient user experience.

# All Promptflow Contribution checklist:
- [X] **The pull request does not introduce [breaking changes].**
- [X] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [X] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [X] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [X] Title of the pull request is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [X] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_core/openai_injector.py [hunk 1]
```diff
@@ -81,22 +111,41 @@ def get_aoai_telemetry_headers() -> dict:
...
+        @functools.wraps(f)
+        async def wrapper(*args, **kwargs):
+            # Inject headers from operation context, overwrite injected header with headers from kwargs.
+            injected_headers = get_aoai_telemetry_headers()
+            original_headers = kwargs.get("headers") if IS_LEGACY_OPENAI else kwargs.get("extra_headers")
+            if original_headers and isinstance(original_headers, dict):
+                injected_headers.update(original_headers)
+            kwargs.update(headers=injected_headers) if IS_LEGACY_OPENAI else kwargs.update(
+                extra_headers=injected_headers
+            )
```
Comment (zzn2) [inline src/promptflow/promptflow/_core/openai_injector.py:125]: Let's extract this part into a subfunction.

Comment (liucheng-ms) [reply src/promptflow/promptflow/_core/openai_injector.py]: Yes. I'd like to do this.

**Review by zzn2 — DISMISSED** [review]

File: src/promptflow/promptflow/_core/openai_injector.py [hunk 2]
```diff
@@ -133,11 +214,27 @@ def inject_openai_api():
...
+    for openai_api_async in available_openai_apis_async():
+        # Due to the issue reported in https://github.com/openai/openai-python/issues/996,
+        # Async interface was not recognized as a coroutine by inspect.iscoroutinefunction.
+        # This means that async injection needs to be handled differently to ensure decorators
+        # behave correctly for asynchronous methods.
+
+        # Check if the create method of the openai_api class has already been modified
+        if hasattr(openai_api_async, "acreate") and not hasattr(openai_api_async.acreate, "_original"):
+            # If not, modify it by calling the inject function with it as an argument
+            openai_api_async.acreate = inject_async(openai_api_async.acreate)
```
Comment (zzn2) [inline src/promptflow/promptflow/_core/openai_injector.py:232]: Add `if IS_LEGACY_OPENAI` to this block?

File: src/promptflow/promptflow/_core/openai_injector.py [hunk 3]
```diff
@@ -81,27 +119,70 @@ def get_aoai_telemetry_headers() -> dict:
...
+                openai_api = getattr(openai, api)
+                if hasattr(openai_api, "acreate"):
+                    yield openai_api
+            except AttributeError:
+                # This is expected for older versions of openai or unsupported APIs.
+                # E.g. ChatCompletion API was introduced in 2023 and requires openai>=0.27.0 to work.
+                # Older versions of openai do not have this API and will raise an AttributeError if we try to use it.
+                pass
+    else:
+        for api in ("AsyncCompletions", "AsyncChat", "AsyncEmbeddings"):
```
Comment (zzn2) [inline src/promptflow/promptflow/_core/openai_injector.py:172]:
```python
(
  (openai.resource.chat, "AsyncCompletions"),
  (openai.resource, "AsyncChat"),
  (openai.resource, "AsyncEmbeddings"),
)
```

File: src/promptflow/promptflow/_core/openai_injector.py (hunk 2 above)
Comment (liucheng-ms) [reply src/promptflow/promptflow/_core/openai_injector.py]: OK. I will add the block

File: src/promptflow/promptflow/_core/openai_injector.py (hunk 3 above)
Comment (liucheng-ms) [reply src/promptflow/promptflow/_core/openai_injector.py]: Updated

File: src/promptflow/promptflow/_core/openai_injector.py [hunk 4]
```diff
@@ -81,63 +119,106 @@ def get_aoai_telemetry_headers() -> dict:
...
+
+        async_apis = (
+            (openai.resources.chat, "AsyncCompletions", "create", inject_async),
+            (openai.resources, "AsyncCompletions", "create", inject_async),
+            (openai.resources, "AsyncEmbeddings", "create", inject_async),
+        )
+
+    # Helper function to check if the API methods exist and yield the valid tuples
+    def check_apis(apis):
+        for module, clz, method, injector in apis:
```
Comment (zzn2) [inline src/promptflow/promptflow/_core/openai_injector.py:199]: `clz` is usually used to indicate an instance of a class. So do the `method`.
I'd prefer to use `class_name`, `method_name` here to indicate they are names (not instances).
On the other hand, `method` is an instance, so it is a good name here.
```suggestion
        for module, class_name, method_name, injector in apis:
```

File: src/promptflow/promptflow/_core/openai_injector.py [hunk 5]
```diff
@@ -133,11 +226,10 @@ def inject_openai_api():
     It stores the original methods as _original attributes of the create methods.
     2. Updates the openai api configs from environment variables.
     """
-    for openai_api in available_openai_apis():
-        # Check if the create method of the openai_api class has already been modified
```
Comment (zzn2) [inline src/promptflow/promptflow/_core/openai_injector.py:137]: Please keep this line of comment as well.

Comment (liucheng-ms) [reply src/promptflow/promptflow/_core/openai_injector.py:137]: OK

**Review by zzn2 — DISMISSED** [review]

File: src/promptflow/promptflow/_core/openai_injector.py [hunk 6]
```diff
@@ -81,50 +121,103 @@ def get_aoai_telemetry_headers() -> dict:
...
+
+
+def inject_sync(f):
+    wrapper_fun = inject_operation_headers((inject_function_sync(["api_key", "headers", "extra_headers"])(f)))
     wrapper_fun._original = f
     return wrapper_fun
 
 
-def available_openai_apis():
+def _openai_api_list(sync=True):
```
Comment (thy09) [inline src/promptflow/promptflow/_core/openai_injector.py:161]: Maybe we can directly return a tuple list? (sync_apis, async_apis)

File: src/promptflow/promptflow/_core/openai_injector.py [hunk 7]
```diff
@@ -81,50 +121,103 @@ def get_aoai_telemetry_headers() -> dict:
...
+        async_apis = (
+            ("openai.resources.chat", "AsyncCompletions", "create"),
+            ("openai.resources", "AsyncCompletions", "create"),
+            ("openai.resources", "AsyncEmbeddings", "create"),
+        )
+
+    yield from sync_apis if sync else async_apis
+
+
+def _generate_api_and_injector(apis, sync=True):
```
Comment (thy09) [inline src/promptflow/promptflow/_core/openai_injector.py:190]: Seems we can do something like this:
```
api_tuples = [(sync_apis, inject_sync), (async_apis, inject_async)]
for apis, injector in api_tuples:
    for module_name, class_name, method_name in apis:
         # Do stuffs
```

Comment (liucheng-ms) [reply src/promptflow/promptflow/_core/openai_injector.py:190]: What's the benefit of doing this?

File: src/promptflow/promptflow/_core/openai_injector.py (hunk 6 above)
Comment (liucheng-ms) [reply src/promptflow/promptflow/_core/openai_injector.py:161]: Seems that user can not tell which is sync and which is async from the returning.

File: src/promptflow/promptflow/_core/openai_injector.py (hunk 7 above)
Comment (thy09) [reply src/promptflow/promptflow/_core/openai_injector.py:190]: I saw a lot of `sync=True`, I think it is not necessary.

Comment (liucheng-ms) [reply src/promptflow/promptflow/_core/openai_injector.py:190]: If sync and async is mixed, we will have to know which is sync which is async.

Comment (liucheng-ms) [reply src/promptflow/promptflow/_core/openai_injector.py]: Updated. Please help review.

File: src/promptflow/promptflow/_core/openai_injector.py (hunk 6 above)
Comment (liucheng-ms) [reply src/promptflow/promptflow/_core/openai_injector.py]: Return a tuple with injector together.

**Review by zzn2 — APPROVED** [review]

**Review by thy09 — APPROVED** [review]

**Pull request merged**